### ISSUE TITLE
Uncaught TypeError: Argument 1 passed to FaaPz\PDO\DatabaseException::__construct() must be of the type string, null given

### ISSUE THREAD

**Issue by nixoncode** [FaaPz/PDO#139]
> php -S 0.0.0.0:8080 -t public
[Wed Oct 28 22:23:19 2020] PHP 7.4.11 Development Server (http://0.0.0.0:8080) started
[Wed Oct 28 22:23:22 2020] 127.0.0.1:60524 Accepted
[Wed Oct 28 22:23:22 2020] PHP Fatal error:  Uncaught TypeError: Argument 1 passed to FaaPz\PDO\DatabaseException::__construct() must be of the type string, null given, called in /Users/nixon/PhpstormProjects/my-blog/vendor/faapz/pdo/src/AbstractStatement.php on line 40 and defined in /Users/nixon/PhpstormProjects/my-blog/vendor/faapz/pdo/src/DatabaseException.php:23
Stack trace:
#0 /Users/nixon/PhpstormProjects/my-blog/vendor/faapz/pdo/src/AbstractStatement.php(40): FaaPz\PDO\DatabaseException->__construct(NULL, 'HY093')
#1 /Users/nixon/PhpstormProjects/my-blog/vendor/faapz/pdo/src/Statement/Update.php(155): FaaPz\PDO\AbstractStatement->execute()
#2 /Users/nixon/PhpstormProjects/my-blog/app/Models/User.php(73): FaaPz\PDO\Statement\Update->execute()
#3 /Users/nixon/PhpstormProjects/my-blog/app/Services/Auth.php(94): App\Models\User->update('id', 6, Array)
#4 /Users/nixon/PhpstormProjects/my-blog/src/dependencies.php(33): App\Services\Auth->user()
#5 [internal function]: {closure}(Object(DI\Container))
#6 /Users/nixon/PhpstormProjects/celco in /Users/nixon/PhpstormProjects/my-blog/vendor/faapz/pdo/src/DatabaseException.php on line 23
[Wed Oct 28 22:23:22 2020] 127.0.0.1:60524 [200]: GET /partners - Uncaught TypeError: Argument 1 passed to FaaPz\PDO\DatabaseException::__construct() must be of the type string, null given, called in /Users/nixon/PhpstormProjects/my-blog/vendor/faapz/pdo/src/AbstractStatement.php on line 40 and defined in /Users/nixon/PhpstormProjects/my-blog/vendor/faapz/pdo/src/DatabaseException.php:23
Stack trace:
#0 /Users/nixon/PhpstormProjects/my-blog/vendor/faapz/pdo/src/AbstractStatement.php(40): FaaPz\PDO\DatabaseException->__construct(NULL, 'HY093')
#1 /Users/nixon/PhpstormProjects/my-blog/vendor/faapz/pdo/src/Statement/Update.php(155): FaaPz\PDO\AbstractStatement->execute()
#2 /Users/nixon/PhpstormProjects/my-blog/app/Models/User.php(73): FaaPz\PDO\Statement\Update->execute()
#3 /Users/nixon/PhpstormProjects/my-blog/app/Services/Auth.php(94): App\Models\User->update('id', 6, Array)
#4 /Users/nixon/PhpstormProjects/my-blog/src/dependencies.php(33): App\Services\Auth->user()
#5 [internal function]: {closure}(Object(DI\Container))
#6 /Users/nixon/PhpstormProjects/celco in /Users/nixon/PhpstormProjects/my-blog/vendor/faapz/pdo/src/DatabaseException.php on line 23
[Wed Oct 28 22:23:22 2020] 127.0.0.1:60524 Closing
[Wed Oct 28 22:25:00 2020] 127.0.0.1:60542 Accepted
[Wed Oct 28 22:25:00 2020] 127.0.0.1:60542 [200]: GET /partners
[Wed Oct 28 22:25:00 2020] 127.0.0.1:60542 Closing

Please help with the error above

**Comment by kwhat** [comment]
How did you construct the Database object? I suspect you inadvertently passed null instead of the DSN.

**Comment by nixoncode** [comment]
I have the database object initialized correctly, other queries are running well

**Comment by kwhat** [comment]
Maybe a basic example of how to duplicate the problem would help me understand what is happening.

**Comment by nixoncode** [comment]
I somehow managed to fix it by using `date()` function instead of Faapz `new Raw('NOW()')`

```php
$this->db->update(['last_seen' => date('Y-m-d H:i:s')])
                ->table('users')
                ->where(new Conditional('id', '=', 1))
                ->execute();
```

The initial query was like this

```php
 $this->db->update(['last_seen' => new Raw('NOW()')])
                ->table('users')
                ->where(new Conditional('id', '=', 1))
                ->execute();
```

how can I use MySQL's now Function without using the PHP date?

**Comment by nixoncode** [comment]
> Maybe a basic example of how to duplicate the problem would help me understand what is happening.

the queries above can help?

**Comment by kwhat** [comment]
Yes that helps, I will need to do a little testing on my side.  I will get back to you shortly.

**Comment by kwhat** [comment]
Hm, so I noticed that you are getting a DatabaseException so I think you are using 2.0.0?  The DatabaseException was removed in 2.0.1 due to some issues with __toString and throwing Exceptions. Is it possible to try with 2.0.1 and provide a stack trace if it crashes?

**Comment by nixoncode** [comment]
![16039518378135421810261179010498](https://user-images.githubusercontent.com/15726312/97532124-ba046000-19c6-11eb-9307-1b4c5efaced8.jpg)
Am already using 2.0.1

**Comment by kwhat** [comment]
Raw clause is broken.

**Comment by kwhat** [comment]
There is also another bug with the `list($state, $code, $message) = $stmt->errorInfo();` code in the `AbstractStatement` i'll take a look at it for 2.1.

**Comment by kwhat** [comment]
Quick question, I presume you are using PDO::ATTR_ERRMODE => PDO::ERRMODE_SILENT or PDO::ERRMODE_WARNING.  How do you check for errors in your code?  How would you prefer to check for errors on `->execute()` without exceptions?  I was thinking though a couple of options for this library: execute could return bool false or return the PDOStatement object where `->errorCode()` could be used.  Let me know what you think would work best.

**Comment by kwhat** [comment]
Fix:

```
diff --git a/src/Statement/Update.php b/src/Statement/Update.php
index 345adde..d34de6e 100644
--- a/src/Statement/Update.php
+++ b/src/Statement/Update.php
@@ -72,7 +72,14 @@ class Update extends AdvancedStatement
      */
     public function getValues(): array
     {
-        $values = array_values($this->pairs);
+        $values = [];
+        foreach ($this->pairs as $value) {
+            if ($value instanceof QueryInterface) {
+                $values = array_merge($values, $value->getValues());
+            } else {
+                $values[] = $value;
+            }
+        }
 
         if ($this->where !== null) {
             $values = array_merge($values, $this->where->getValues());
```

**Comment by nixoncode** [comment]
> Quick question, I presume you are using PDO::ATTR_ERRMODE => PDO::ERRMODE_SILENT or PDO::ERRMODE_WARNING. How do you check for errors in your code? How would you prefer to check for errors on `->execute()` without exceptions? I was thinking though a couple of options for this library: execute could return bool false or return the PDOStatement object where `->errorCode()` could be used. Let me know what you think would work best.

Am using the Default, I checked the code and saw `PDO::ERRMODE_EXCEPTION` was already set. I always do this when using raw PDO. 

Am currently handling errors by wrapping everything in a try-catch and listen for errors on `FaaPz\PDO\DatabaseException` I think that works just fine, in the past I would do \PDOException. I think exceptions is the best way to go since PDO does that as well.

**Comment by nixoncode** [comment]
> There is also another bug with the `list($state, $code, $message) = $stmt->errorInfo();` code in the `AbstractStatement` i'll take a look at it for 2.1.

Yes, at times null is returned for the message. I don't know why. I remember making the message nullable on the `DatabaseException` class

```php
public function __construct(?string $message = '', string $code = 'database_error', Exception $previous = null)
    {
        parent::__construct($mes sage, 0, $previous);
        $this->code = $code;
    }
```

**Comment by kwhat** [comment]
I published 2.1 which fixes your issue.  I think it has something to do with the mysql pdo driver not having a msg for that particular error code.  Either way, DatabaseException has gone away and I simplified a few things to make it easier to work with. Let me know if you run into any other issues.
